README.md
## Null Safety support
Fork packages [flutter_aes_ecb_pkcs5](https://pub.dartlang.org/packages/flutter_aes_ecb_pkcs5) 

Solve the null safety support problem


## AES 128 pkcs5 only support
* Provide generation of AES keys
* Provide encrypt for AES keys  
* Provide decrypt for AES keys  

* Android uses java, iOS uses OC development


## Usage

Use this package as a library
1. Depend on it
Add this to your package's pubspec.yaml file:

```
dependencies:
  flutter_aes_ecb_pkcs5_fork: ^0.1.4
```

2. Install it You can install packages from the command line:

with Flutter:

```
$ flutter pub get
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

3. Import it Now in your Dart code, you can use:

```
import 'package:flutter_aes_ecb_pkcs5_fork/flutter_aes_ecb_pkcs5_fork.dart';
```



#### AES Use Example



```dart
import 'package:flutter_aes_ecb_pkcs5_fork/flutter_aes_ecb_pkcs5_fork.dart';

void main() async {
  var data = "{\"username\":\"helloword\"}";

  //generate a 16-byte random key
  var key = await FlutterAesEcbPkcs5.generateDesKey(128);

  print(key);
  //encrypt
  var encryptText = await FlutterAesEcbPkcs5.encryptString(data, key);

  print(encryptText);
  //decrypt
  var decryptText = await FlutterAesEcbPkcs5.decryptString(encryptText, key);

  print(decryptText);
}

```

#### Run Example Result
```
data:{"username":"helloword"}
create key:DA8FD2693B652D587EEE51E7ADD9DBB8
encryptText :851544E224F6C1C7F560AB82889DE0DF66EEB6762AB836F4EE13E31100F0EA16
decryptText :{"username":"helloword"}
```

#### License

The MIT License(http://opensource.org/licenses/MIT)

Please feel free to use and contribute to the development.




CHANGELOG.md
## 0.0.1

* first create aes ecb pkcs5 .

## 0.0.2

* update readme and Code comment.

## 0.1.0
* update readme and Code comment and description.
* its release.

## 0.1.1
* change description. Null Safety support.

## 0.1.2
* migrate it to Android embedding v2.

## 0.1.3
* format dart code.

## 0.1.4
* support ios.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_aes_ecb_pkcs5_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_aes_ecb_pkcs5_fork_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_aes_ecb_pkcs5_fork/flutter_aes_ecb_pkcs5_fork.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_aes_ecb_pkcs5');

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterAesEcbPkcs5.platformVersion, '42');
  });
}


lib/flutter_aes_ecb_pkcs5_fork.dart
import 'dart:async';

import 'package:flutter/services.dart';

/// AES ECB PKCS5 encrypt & decrypt.
class FlutterAesEcbPkcs5 {
  static const MethodChannel _channel =
      const MethodChannel('flutter_aes_ecb_pkcs5');

  /// Get device system version.
  ///
  /// * If it is a Android device, like 'Android 9.0'.
  /// * If it is a IOS device, like 'IOS 11.2'.
  static Future<String> get platformVersion async {
    final String version =
        await _channel.invokeMethod('getPlatformVersion') ?? "";
    return version;
  }

  /// Used to generate a specific length of AES key with the given [length].
  ///
  /// If it is a 128-bit key, enter 128 and the returned key will be encoded by Hex.
  /// [value] and [key] shoudn't be null.
  static Future<String> generateDesKey(int length) async {
    try {
      final String result =
          await _channel.invokeMethod('generateDesKey', {"length": length});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }

  /// Encrypts the [txt] with the given [key].
  ///
  /// The result of the encryption will be encoded by HEX.
  /// [txt] and [key] shoudn't be null.
  static Future<String> encryptString(String txt, String key) async {
    try {
      final String result =
          await _channel.invokeMethod('encrypt', {"input": txt, "key": key});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }

  /// Decrypts the [txt] with the given [key].
  ///
  /// The content [txt] that needs to be decrypted must be HEX encoded，key also.
  /// [txt] and [key] shoudn't be null.
  static Future<String> decryptString(String txt, String key) async {
    try {
      final String result =
          await _channel.invokeMethod('decrypt', {"input": txt, "key": key});
      return result;
    } on PlatformException catch (e) {
      throw "Failed to get string encoded: '${e.message}'.";
    }
  }
}


example/ios/Flutter/flutter_export_environment.sh
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export "FLUTTER_ROOT=/Users/xx/soft/flutter"
export "FLUTTER_APPLICATION_PATH=/Users/xx/Documents/1project/flutter/flutter_aes_ecb_pkcs5_fork/example"
export "COCOAPODS_PARALLEL_CODE_SIGN=true"
export "FLUTTER_TARGET=/Users/xx/Documents/1project/flutter/flutter_aes_ecb_pkcs5_fork/example/lib/main.dart"
export "FLUTTER_BUILD_DIR=build"
export "FLUTTER_BUILD_NAME=1.0.0"
export "FLUTTER_BUILD_NUMBER=1"
export "DART_DEFINES=Zmx1dHRlci5pbnNwZWN0b3Iuc3RydWN0dXJlZEVycm9ycz10cnVl,RkxVVFRFUl9XRUJfQVVUT19ERVRFQ1Q9dHJ1ZQ=="
export "DART_OBFUSCATION=false"
export "TRACK_WIDGET_CREATION=true"
export "TREE_SHAKE_ICONS=false"
export "PACKAGE_CONFIG=/Users/xx/Documents/1project/flutter/flutter_aes_ecb_pkcs5_fork/example/.dart_tool/package_config.json"


android/src/main/java/com/smallbuer/flutter_aes_ecb_pkcs5/FlutterAesEcbPkcs5Plugin.java
package com.smallbuer.flutter_aes_ecb_pkcs5;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** FlutterAesEcbPkcs5Plugin */
public class FlutterAesEcbPkcs5Plugin implements  FlutterPlugin, MethodCallHandler {

  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_aes_ecb_pkcs5");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
  

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    }else if (call.method.equals("generateDesKey")) {
      int length = call.argument("length");
      if (length != 0 ) {
        try {
          String encoded = new AesSecurity().generateDesKey(length);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "generateDesKey failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "generateDesKey failure.", null);
      }

    }else if(call.method.equals("encrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String encoded = new AesSecurity().encrypt(text, publicKey);
          result.success(encoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Encrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Encrypt failure.", null);
      }
    }else if(call.method.equals("decrypt")){

      String text = call.argument("input");
      String publicKey = call.argument("key");
      if (text != null && publicKey != null) {
        try {
          String decoded = new AesSecurity().decrypt(text, publicKey);
          result.success(decoded);
        } catch (Exception e) {
          e.printStackTrace();
          result.error("UNAVAILABLE", "Decrypt failure.", null);
        }
      }else{
        result.error("NULL INPUT STRING", "Decrypt failure.", null);
      }
    } else {
      result.notImplemented();
    }
  }

}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_aes_ecb_pkcs5_fork/flutter_aes_ecb_pkcs5_fork.dart';

void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String pkcs5Progress;

    //string format
    var data = "{\"username\":\"helloword\"}";

    //create 16 byte random key
    var key = await FlutterAesEcbPkcs5.generateDesKey(128);

    //encrypt
    var encryptText = await FlutterAesEcbPkcs5.encryptString(data, key);

    //decrypt
    var decryptText = await FlutterAesEcbPkcs5.decryptString(encryptText, key);

    pkcs5Progress = "data:" +
        data +
        "\n" +
        "create key:" +
        key +
        "\n" +
        "encryptText :" +
        encryptText +
        "\n" +
        "decryptText :" +
        decryptText +
        "\n";

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = pkcs5Progress;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('flutter_aes_ecb_pkcs5'),
        ),
        body: Center(
          child: Text('pkcs5Progress:\n $_platformVersion\n'),
        ),
      ),
    );
  }
}
